### PR TITLE
[SofaCore] Reorganize ObjectFactory so it becomes possible to refactor it in a smooth way

### PR DESCRIPTION
**Pull request by damienmarchal**
Refactoring ObjectFactory without breaking to much SOFA is rather complex. 
To ease in the process I split the ObjectFactory class in three entity.
 
sofa::core::future::ObjectFactory and sofa::core::future::ObjectFactoryInstance are the new versions while sofa::core::ObjectFactory is a compatibility layer exposing the old API for transitionning.  

For the moment, 100% of the old API is reproduced and emits deprecation warning. 
As the code change are rather important it is easier to read the source code instead of the diff. 

sofa::core::future::ObjectFactory is implementing the factory mecanisme without any static in it. 
sofa::core::future::ObjectFactoryInstance is a single-instance factory  accessible through static method. 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by guparan** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h [hunk 1]
```diff
@@ -19,121 +19,77 @@
...
+#include <sofa/core/future/ObjectFactory.h>
+#include <sofa/core/future/ObjectFactoryInstance.h>
 
-namespace sofa
+namespace sofa::core
 {
 
-namespace core
-{
+// inject in the current namespace the
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h]: the ?

File: SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h [hunk 2]
```diff
@@ -19,121 +19,77 @@
...
-    /// Main class registry
-    ClassEntryMap registry;
-    OnCreateCallback m_callbackOnCreate ;
+    using Creator = sofa::core::future::ObjectFactory::Creator;
+    using CreatorMap = sofa::core::future::ObjectFactory::CreatorMap;
 
-public:
+    ObjectFactory(sofa::core::future::ObjectFactory* newfactory_){ newfactory=newfactory_; }
+
+    SOFA_ATTRIBUTE_DEPRECATED__OBJECTFACTORY("short name has been removed. Use sofa::helper::NameDecoder::sortName instead.")
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h]:
```suggestion
    SOFA_ATTRIBUTE_DEPRECATED__OBJECTFACTORY("Use sofa::helper::NameDecoder::sortName instead.")
```

File: SofaKernel/modules/Sofa.Config/src/sofa/config.h.in [hunk 3]
```diff
@@ -305,6 +305,11 @@ class DeprecatedAndRemoved {};
         "v21.12 (PR#2292)", "v22.06", \
         "Use the Messaging API instead of using SofaOStream and sout/serr/sendl.")
 
+#define SOFA_ATTRIBUTE_DEPRECATED__OBJECTFACTORY(msg) \
+    SOFA_ATTRIBUTE_DEPRECATED( \
+        "v21.12 (PR#xxx)", "v22.12", \
+        "A new API has been proposed for ObjectFactory. Please update your code to use this one." msg)
+
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Config/src/sofa/config.h.in]: Should be moved to SofaKernel/modules/SofaCore/src/sofa/core/config.h.in

**Comment by damienmarchal** [comment]
@hugtalbot  
- it is not related to #1087, 
- it is more to prepare the evolution from the class that mix in a single object a static/singleton part of the object and a part that shouldn't be one. It was often said in discussion that such design was highly questionnable. 
- i prefer to have such move before implementing the import/namespace in the factory demonstrated in https://github.com/sofa-framework/sofa/pull/2512
  
About the use of the "future" namespace I use it to clearly show that there is two "ObjectFactory". The one with the old API (and it fully backward compatible) and the one with the new API. I could of course have mixed everything  (the old and the new API) in a single one object but I found it quite convenient to clearly see the objects future::ObjectFatory  and ObjectFactoryInstance in their final state and well separated from the "compatbility" layer in ObjectFactory.   
I could have use a different class name for future::ObjectFactory like NewObjectFactory but this is would have been behaving exactly like having a namespace glued to the class name (actually I see this a namespace hidden in the classname so people don't notice it is one ;))
I also consider using the approach of SofaNG to make deprecation but this only work if we move the new version of the object to be in a different namespace that the one we want to deprecate. In the present case I wasn't sure we really want to move sofa::core;;ObjectFactory somewhere less, if you find an appropriate suggestion i would be very welcome as I would found that easier than using the "extra" namespace. 
In NG we often do such a thing but this is done like that:
1) move the new API to a new namespace  (eg: sofa::type::)
2) make a deprecation layer                         (eg: sofa::defaulttypes::)
In the ObjectFactory cas I'm not sure there is a need to have the new api moved in a new namespace (I see no problem in having it in sofa::core)  

EDIT: and "future" this is nothing related to master branch. As being in master branch doesn't tells how deprecated API should be implemented on top of the new one during the transitional period.

File: SofaKernel/modules/Sofa.Config/src/sofa/config.h.in (hunk 3 above)
Comment (damienmarchal) [reply SofaKernel/modules/Sofa.Config/src/sofa/config.h.in]: Right.

NB: if this is the guideline we probably should we move all the deprecation that are in Sofa.Config/config.h.in into their corresponding package  ?

Comment (guparan) [reply SofaKernel/modules/Sofa.Config/src/sofa/config.h.in]: Yes we should :+1:

File: SofaKernel/modules/SofaHelper/src/sofa/helper/StringUtils.h [hunk 4]
```diff
@@ -53,6 +55,25 @@ std::string join(std::vector<T>& elems, S& delim) {
     }
     return ss.str();
 }
+
+///@brief Join a std::vector into a single string, separated by the provided delimiter and a provided to string converion function
```
Comment (alxbilger) [inline SofaKernel/modules/SofaHelper/src/sofa/helper/StringUtils.h]:
```suggestion
///@brief Join a std::vector into a single string, separated by the provided delimiter and a provided to string conversion function
```

File: SofaKernel/modules/SofaCore/src/sofa/core/future/ObjectFactoryInstance.h [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    const char* getHeaderFileLocation() override
+    {
+        return RealObject::HeaderFileLocation();
+    }
+};
+
+/**
+ *  \brief Helper class used to register a class in the ObjectFactory.
+ *
+ *  This class accumulate information about a given class, as well as creators
```
Comment (alxbilger) [inline SofaKernel/modules/SofaCore/src/sofa/core/future/ObjectFactoryInstance.h]:
```suggestion
 *  This class accumulates information about a given class, as well as creators
```

File: SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h [hunk 6]
```diff
@@ -19,174 +19,99 @@
...
```
Comment (alxbilger) [inline SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h]:
```suggestion
    SOFA_ATTRIBUTE_DEPRECATED__OBJECTFACTORY("short name has been removed. Use sofa::helper::NameDecoder::shortName instead.")
```

**Pull request closed without merging**